rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,869 @@
+The user will be able to rollback to a previous policy version by updating the `Policy` CRD (via a spec field or an annotation, TBD).
+The controller will detect the change and use the `PolicyRevision` CRD to update the Policy CRD with the previous configuration.
+
+### Garbage Collection
+
+When a `Policy` is deleted, the controller will also delete all associated `PolicyRevisions`.
+Using the `OwnerReference` field is not feasible in this case, as the `PolicyRevision` may reside in a different namespace than the `AdmissionPolicy` or `AdmissionPolicyGroup`,
+or it could be versioning a `ClusterAdmissionPolicy` or `ClusterAdmissionPolicyGroup`, which are cluster-scoped resources.
+
+We will also need to implement a garbage collection mechanism to remove old precompiled modules from the shared cache.
+This will prevent the cache from growing indefinitely and consuming unnecessary resources.
+
+## Policy Lifecycle
+
+Given the concepts described above, the policy lifecycle will be as follows:
+
+### Policy Creation
+
+1. The user creates a new `Policy`.
+2. The Kubewarden controller creates a new `PolicyRevision` resource with the policy content and sets the `enabled` field to `true`.
+   It also sets the `policyGeneration` field to the value of `metadata.generation` from the `Policy` CRD and adds the `Scheduled` status condition set to `False` to both the `Policy` and `PolicyRevision` CRDs.
+3. The new `PolicyRevision` triggers the leader reconciler. The Policy Server leader sets the status condition of type `Scheduled` of the `PolicyRevision` to `True`.
+   - If Policy Server specified in the `Policy` CRD is not running, the status condition of type `Scheduled` will be set to `False` with the appropriate reason and message.
+4. The Kubewarden controller propagates the status condition to the `Policy` CRD.
+5. The Policy Server leader pulls the policy from the registry, precompiles the policy, and stores it in the shared cache.
+6. The Policy Server leader validates the policy settings.
+7. The Policy Server leader changes the status condition of type `Initialized` of the `PolicyRevision` to `True`.
+   - If an error occurred in the previous steps, the status condition of type `Initialized` will be set to `False` with the appropriate reason and message.
+8. The Kubewarden controller propagates the status condition to the `Policy` CRD.
+9. The replica reconciler of all the replicas including the leader loads the policy in the evaluation environment.
+10. Every replica will report the status of the policy to the `PolicyRevision` CRD, setting the status condition of type `Ready` to `True`.
+    - If an error occurs, the status condition of type `Ready` will be set to `False` with the appropriate reason and message.
+11. The `Ready` status is propagated to the `Policy` CRD.
+12. When all the replicas have set the `Ready` status to true, the Kubewarden controller creates a Webhook configuration pointing to the `PolicyRevision` generation.
```

Comment:
Can we ensure we are eventually consistent if the replica numbers shuffle, grow or diminish?